Add dfa-size-limit and regex-size-limit arguments

tiehuis opened the pull request:
See #362.

 - The macro is not really to my taste but I wanted to see if you were happy with it first or had any ideas otherwise. I did try creating a generic function with the required trait bounds but the generic multiplication was the stopping point here for me. Also, if using a `saturating_mul` this is further pronounced since there is currently no trait in std which these are bound to. The `Saturating` trait on the [num](https://github.com/rust-num/num) also only applies for addition to.

- Removed the straight-forward multiplications and replaced with saturating versions. This avoids the potential overflow. Added a test case for this also to ensure it is handled properly. Since these arguments are always some memory bound (as per the suffix) a saturating limit makes the most sense. That being said, it is very (very) unlikely for the current width types.

kpp commented on src/args.rs:
```diff
@@ -312,6 +312,48 @@ impl Args {
+// correctly constraining the output $ty to work for a number of integers.
+// Since there are limited numbers of types that currently need this function
+// we perform the type monomorphization ourselves.
+macro_rules! human_readable_size_arg_impl {
```
You can split this macro into 2 functions: validator and value converter. See [an example of validator](https://github.com/kbknapp/clap-validators/blob/master/src/num/mod.rs)

BurntSushi replied:
I'm not sure this macro is necessary, and I'd really like to find a way to get rid of it. Can it be written to specifically return a `u64` and then just convert the result to a `usize`? (That conversion should return an error if the `u64` is greater than [`std::usize::MAX`](https://doc.rust-lang.org/std/usize/constant.MAX.html).)

BurntSushi commented on src/app.rs:
```diff
@@ -340,6 +344,13 @@ lazy_static! {
+        doc!(h, "dfa-size-limit",
+             "The upper size limit of the generated dfa.",
+             "The upper size limit of the generated dfa. The default limit is \
+              10Mb. This should only be changed on very large regex inputs \
```
I think this should say `10M` instead of `10Mb`. (Even so, the `Mb` suffix means "megabit", which is otherwise wrong. :-))

BurntSushi commented on src/app.rs:
```diff
@@ -454,6 +465,11 @@ lazy_static! {
+        doc!(h, "regex-size-limit",
+             "The upper size limit of the compiled regex.",
+             "The upper size limit of the compiled regex. The default limit \
+              is 10Mb. \n\nThe argument accepts the same size suffixes as \
```
Same here. `10M` instead of `10Mb`.

BurntSushi commented:
Could you also change the commit message to say, `Fixes #362` so that github properly connects the PR with the issue? Thanks.

tiehuis commented:
Made fixes as per your suggestions and squashed into a single commit.

I've also added overflow detection instead of the saturation as you argued. Figured that it was better to get it as intended right now rather than leaving it for later.

Added an extra test specifically for 32-bit systems for when `usize < u64`.

BurntSushi commented:
@tiehuis Thanks again for working on this! I merged this PR in https://github.com/BurntSushi/ripgrep/commit/66efbad871620fe2dbe675438fdc8d9922e72826. I ended up making a few cosmetic changes to make the code more consistent with the style of the rest of the code. Nice work! :-)

kpp commented:
@BurntSushi stop closing pull requests. There is a button to merge the code with any message you like.

BurntSushi commented:
@kpp If you have a problem with the way I'm running the project and can suggest a better alternative, then please file an issue.
